Uniswap Guide Added (#415)
subquery/documentation: docs/quickstart/quickstart_chains/ethereum-uniswap.md

# Ethereum Quick Start - Uniswap (Medium)

## Goals

> Within the domain of DeFi, the objective of indexing Uniswap entities is both crucial and versatile. Essentially, indexing serves as the foundation for transparency and ease of use in the Uniswap system. By systematically organizing tokens, liquidity pools, transactions, and other essential information, indexing provides users with a quick and efficient means to search, find, and analyze data.

Reviewer: What does this mean?

Reviewer 2: Makes sense to me

Reviewer: it's way too over the top in my view

> The objective of this article is to offer a detailed, step-by-step guide on setting up a Subquery indexer for Uniswap data indexing. We will comprehensively cover the necessary configurations and delve into the intricacies of the underlying logic.

Reviewer: ```suggestion
The objective of this article is to offer a detailed, step-by-step guide on setting up a SubQuery indexer for Uniswap data indexing. We will comprehensively cover the necessary configurations and delve into the intricacies of the underlying logic.
```

> Previously, in the [1. Create a New Project](../quickstart.md) section, you must have noted [3 key files](../quickstart.md#_3-make-changes-to-your-project). Let's begin updating them one by one.

Reviewer 2: @bgodlin 

I like all the changes. I have just one question related with the guide dependencies, previous steps reader needs to take to follow it. Shall we mention which project should this person initialise via SQ CLI?

Author: Added as part of https://github.com/subquery/documentation/pull/415/commits/b2000991d87bd155b782b0eb540b2c5b1e6ec632

## Setting Up the Indexer

In this Uniswap indexing project, our main focus is on configuring the indexer to exclusively capture logs generated by three specific types of Uniswap smart contracts:

1. **UniswapV3Factory** (contract address: `0x1F98431c8aD98523631AE4a59f267346ea31F984`): This contract is responsible for creating all the pool smart contracts.

2. **Smart Contracts for Individual Pools**: These contracts represent individual liquidity pools.

3. **NonfungiblePositionManager** (contract address: `0xC36442b4a4522E871399CD717aBDD847Ab11FE88`): This contract is instrumental in producing liquidity positions, which are implemented as NFTs. This functionality enables additional use-cases, including the transfer of liquidity positions.

To gain a deeper understanding of how these core mechanisms work, you can refer to the official [Uniswap documentation](https://docs.uniswap.org/contracts/v3/reference/deployments).

::: warning

> We suggest starting with the [Ethereum Gravatar example](./ethereum-gravatar). The ENS project is a lot more complicated and introduces some more advanced concepts

Reviewer 2: Do you mean - The Ethereum Uniswap project is more complicated?  @bgodlin

Author: > Do you mean - The Ethereum Uniswap project is more complicated? @bgodlin

fixed as part of https://github.com/subquery/documentation/pull/415/commits/4d14525dd66b25bcc2e72c07f85e0c13969b1106

::: tip Note
Throughout this mapping and those that follow, numerous utility functions are employed to process the data. In this specific example, these utility functions are stored in the `utils` directory. If you're interested in understanding how they work, you can refer to the [final code](https://github.com/subquery/ethereum-subql-starter/tree/main/Ethereum/ethereum-uniswap-v3).
:::

3. **Data Persistence**: To ensure the collected data persists, the function saves the modifications made to the `Token`, `Pool`, and `Factory` objects. This typically entails storing the data in a database or data store.

🎉 At this point, you have successfully crafted the handling logic for the factory smart contract and populated queryable entities like `Token`, `Pool`, and `Factory`. This means you can now proceed to the [building process](#build-your-project) to test the indexer's functionality up to this point.

### Pool Smart Contracts

As we discussed in the introduction of [Configuring the Indexer](#configuring-the-indexer), a new contract is created by the [factory contract](#uniswapv3factory) for each newly created pool.

#### 1. Configuring the Manifest File

> Additionally, to handle situations where a contract factory generates fresh contract instances for each new pool, we should utilize [dynamic data sources](../../build/dynamicdatasources.md). Consequently, the following adjustments need to be made in the manifest:

Reviewer: ```suggestion
The contract factory generates fresh contract instances for each new pool, therefore we use [dynamic data sources](../../build/dynamicdatasources.md) to create indexers for each new contract:
```

Reviewer: Shorter and simpler - less complex language is better

However, visit the [Running SubQuery Locally](../../run_publish/run.md) to get more information on the file and the settings.

Run the following command under the project directory:

::: code-tabs
@tab:active yarn

```shell
yarn start:docker
```

@tab npm

```shell
npm run-script start:docker
```

:::

::: tip Note
It may take a few minutes to download the required images and start the various nodes and Postgres databases.
:::

## Query your Project

Next, let's query our project. Follow these three simple steps to query your SubQuery project:

1. Open your browser and head to `http://localhost:3000`.

2. You will see a GraphQL playground in the browser and the schemas which are ready to query.

3. Find the _Docs_ tab on the right side of the playground which should open a documentation drawer. This documentation is automatically generated and it helps you find what entities and methods you can query.

Try the following queries to understand how it works for your new SubQuery starter project. Don’t forget to learn more about the [GraphQL Query language](../../run_publish/query.md).

> ### Pools

Reviewer: Perhaps use the expand markdown component for each example request + response combination to save space here `::: detail xxx`

Author: Added as part of https://github.com/subquery/documentation/pull/415/commits/a83c0df3f088c106285121d2a0f1cec76e535045

# Ethereum Quick Start - Uniswap (Medium)

## Goals

> This project serves as an excellent foundation for initiating your Ethereum SubQuery project. It meticulously indexes various Uniswap entities, including Swaps, Pool operations, Price ticks, Positions, and additionally computes statistical metrics.

Reviewer 2: Minor thing. I'm not sure if this is the best opening line and if this project is actually this great foundation to initialise eth SQ project.  I think I would rather start with what the project indexes, what are the goals.

> ::: warning

Reviewer 2: ```suggestion
::: warning Important
```

We suggest starting with the [Ethereum Gravatar example](./ethereum-gravatar). The Uniswap project is a lot more complicated and introduces some more advanced concepts.
:::

> Now, let's move forward and fork the example code for this project from [here](https://github.com/subquery/ethereum-subql-starter/tree/main/Ethereum/ethereum-uniswap-v3)

Reviewer 2: @bgodlin Why not to use the approach of initialising a boilerplate project and then updating all the files step by step? And only sharing the final code here not asking to fork final version of the project.

Author: I contemplated starting a boilerplate project and creating a step-by-step guide for updating files, but I ultimately chose not to proceed with it. My decision was based on the desire to maintain consistency with the format used in all other published quickstart guides. Now, the question is whether we should introduce a new format for this specific page and subsequently update the format for all other guides, or should we stick with the proposed structure for this article and then implement a new format for all quickstart guides in a single comprehensive pull request?

Reviewer 2: But other published starter guides follow exactly this approach (starter boilerplate + changes to the project with a step by step guide)...?

dataSources:
  - kind: ethereum/Runtime
    startBlock: 12369621
    options:
      # Must be a key of assets
      abi: Factory
      address: "0x1F98431c8aD98523631AE4a59f267346ea31F984"
    assets:
      Factory:
        file: "./abis/Factory.json"
      ERC20:
        file: "./abis/ERC20.json"
      ERC20SymbolBytes:
        file: "./abis/ERC20SymbolBytes.json"
      ERC20NameBytes:
        file: "./abis/ERC20NameBytes.json"
      Pool:
        file: "./abis/pool.json"
    mapping: ...
    # ethereum/contract
  - kind: ethereum/Runtime
    startBlock: 12369651
    options:
      abi: NonfungiblePositionManager
      address: "0xC36442b4a4522E871399CD717aBDD847Ab11FE88"
    assets:
      NonfungiblePositionManager:
        file: ./abis/NonfungiblePositionManager.json
      Pool:
        file: ./abis/pool.json
      Factory:
        file: ./abis/factory.json
      ERC20:
        file: ./abis/ERC20.json
    mapping: ...
templates:
  - name: Pool
    kind: ethereum/Runtime
    options:
      abi: Pool
    assets:
      Pool:
        file: "./abis/pool.json"
      Factory:
        file: "./abis/factory.json"
      ERC20:
        file: "./abis/ERC20.json"
    mapping: ...
```

> 

Reviewer 2: We should also add a paragraph or two explaining what's happening here.

> 

Reviewer 2: ```suggestion
Check out our [Manifest File](https://github.com/subquery/documentation/blob/master/docs/build/manifest/ethereum.md) documentation to get more information about the Project Manifest (project.yaml) file.
```

Author: I think it would be more beneficial to use a link to a local page instead of one that directs to markdown hosted on Github. This is the consistent formatting used in all instances.

Reviewer 2: Ok, please adjust

Check out our [Manifest File](../../build/manifest/ethereum.md) documentation to get more information about the Project Manifest (`project.yaml`) file.

> ## 2. Update Your GraphQL Schema File

Reviewer 2: We are not updating anything here, just going through the code

> The `schema.graphql` file determines the shape of your data from SubQuery due to the mechanism of the GraphQL query language. Hence, updating the GraphQL Schema file is the perfect place to start. It allows you to define your end goal right at the start.

Reviewer 2: If we are not updating schema.graphl this paragraph is not relevant

> Within this project, you will observe the presence of 22 GraphQL entities, each interconnected through numerous foreign key relationships. Consider, for instance, the relationship between the `Pool` and `Token` entities. Importantly, these relationships encompass not only one-to-many connections but also extend to encompass many-to-many associations.

Reviewer 2: I would show Pool and Token entities in the graphql shell

Reviewer 2: Also we can consider adding a link to Hero Course part dedicated to relationships https://academy.subquery.network/academy/herocourse/module3.html#lesson-1-one-to-many-entities

import { Bundle, Burn, Factory, Mint, Pool, Swap, Tick, Token } from "../types";
```

As you're creating a new Etheruem based project, this command will also generate ABI types and save them into `src/types` using the `npx typechain --target=ethers-v5` command, allowing you to bind these contracts to specific addresses in the mappings and call read-only contract methods against the block being processed.

It will also generate a class for every contract event to provide easy access to event parameters, as well as the block and transaction the event originated from. Read about how this is done in [EVM Codegen from ABIs](../../build/introduction.md#evm-codegen-from-abis).

All of these types are written to `src/typs/abi-interfaces` and `src/typs/contracts` directories. In the example Gravatar SubQuery project, you would import these types like so.

```ts
// Import event types from the registry contract ABI
import {
  InitializeEvent,
  MintEvent,
  BurnEvent,
  SwapEvent,
  FlashEvent,
} from "../types/contracts/Pool";
```

Check out the [GraphQL Schema](../../build/graphql.md) documentation to get in-depth information on `schema.graphql` file.

> ## 3. Add a Mapping Function

Reviewer 2: If we are not adding anything to the file, we should change headers

Mapping functions define how chain data is transformed into the optimised GraphQL entities that we previously defined in the `schema.graphql` file.

> They function in a manner akin to SubGraphs, and you can observe in this Uniswap project that they are distributed across three distinct files, along with the inclusion of various utility files.

Reviewer 2: I wouldn't explain how mapping functions work only by referencing to Subgraphs

Check out our [Mappings](../../build/mapping/ethereum.md) documentation to get more information on mapping functions.

> 

Reviewer 2: Also, no examples of mapping functions? We could also add some explanation of what is happening in each function or at least the one we showed as an example in the shell.
